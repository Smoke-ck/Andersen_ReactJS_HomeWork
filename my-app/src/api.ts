export type IToDos = {
    title: string,
    completed: boolean,
    id: string | number,
    favorite: boolean,
}

const API_URL = 'http://localhost:3000/todos';

export const getTodosApi = async (): Promise<IToDos[]> => {
    const responce = await fetch(
        API_URL, {
        method: 'GET'
    }
    )
    return responce.json();
}

export const deleteTodoApi = async (id: number | string): Promise<IToDos> => {
    const responce = await fetch(
        `${API_URL}/${id}`, {
        method: 'DELETE'
    }
    )
    return responce.json()
}

export const createTodoApi = async (title: IToDos): Promise<IToDos> => {
    const responce = await fetch(
        API_URL, {
        method: 'POST',
        body: JSON.stringify({ title: title, completed: false, favorite: false }),
        headers: { 'Content-Type': 'application/json', },
    }
    )
    return responce.json()
}

export const updateTodoApi = async (id: string | number, todo: IToDos): Promise<IToDos> => {
    const responce = await fetch(
        `${API_URL}/${id}`, {
        method: 'PUT',
        body: JSON.stringify({ title: todo.title, completed: todo.completed, favorite: todo.favorite }),
        headers: { 'Content-Type': 'application/json', },
    }
    )
    return responce.json()
}

