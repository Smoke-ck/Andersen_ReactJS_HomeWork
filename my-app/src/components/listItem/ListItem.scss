.listItem__wrapper {
    position: relative;
    word-break: break-word;
}
.listItem {
    background-color:var(--background--color-card);
    margin-bottom: 10px;
    padding: 30px;
    display: flex;
    justify-content: space-between;
    border-radius: 10px;
    &:hover {box-shadow:0 0 10px white;}
    &__title {
        font-size: 25px;
        margin-bottom: 10px;
        color:var(--font-color);
        font-weight: 400;  
    }
    &__wrapper__title {
        position: relative;
    }
    &__button--favorite{
        color:black;
        cursor: pointer;
    }
    &__button--notFavorite {
        color:#ff5500;
        cursor: pointer;
    }
    &__menu {
        cursor: pointer;
        &--title {
            font-family: inherit;
            font-size: 30px;
            position: absolute;
            bottom: 5px;
            left: 30px;
            cursor: pointer;
            &:hover{color:gold}
        }
        &--button {
            margin-top: 10px;
            cursor: pointer;
            vertical-align: top;
            padding: 7px 20px;
            font-family: inherit;
            font-size: 12px;
            text-align: center;
            text-decoration: none;   
            outline: none;
            border: none;
            border-radius: 5px;
            margin-right: 5px;
            &:hover {
                box-shadow:0 0 5px black;
                background-color: gray;
                color: gold;
            }
        }
    }
}
.listItemdone {
    background-color: var(--background--color-done);
    margin-bottom: 10px;
    padding: 30px;
    display: flex;
    justify-content: space-between;
    border-radius: 10px;
    &:hover {box-shadow:0 0 10px white;}
}
.deleteMenu {
    display: flex;
    flex-direction: column;
    align-items: space-around;
    &__title{
        font-family: inherit;
        font-size: 25px;
        color:#63696a; 
        margin-bottom: 10px;
        &--todo {
            font-family: inherit;
            font-size: 20px;
            color:#95251d; 
            margin-bottom: 10px;
        }   
    }
    &__button {
        background-color:#a1a1a1;
        border: none;
        border-radius: 5px;
        color: white;
        padding: 5px 20px;
        text-align: center;
        font-size: 16px;
        margin: 4px 2px;
        opacity: 0.6;
        transition: 0.3s;
        display: inline-block;
        text-decoration: none;
        cursor: pointer;
        font-weight: 600;
        &:hover {opacity: 1;}
        &--cancel {
            color:red;
        }
        &--ok {
            color:#00aa00;
        }
        &--close {
            width: 50px;
            color: red;
            align-self:flex-end;
        }  
    }
}