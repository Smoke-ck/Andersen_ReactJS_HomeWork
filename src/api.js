const API_URL = 'http://localhost:3000/products';

export function getProducts() {
    return request();
}

export function getProduct (id) {
    return request(
        {
            method: 'GET',
        },
        id
    );
}
export function deleteProduct(id) {
    return request(
        {
            method: 'DELETE',
        },
        id
    );
}

export function createProduct(todo) {
    request({
        method: 'POST',
        body: JSON.stringify(todo),
    });
}

export function updateProduct(todo) {
    request(
        {
            method: 'PUT',
            body: JSON.stringify(todo),
        },
        todo.id
    );
}

function request(config = {}, id = '') {
    config = {
        method: 'GET',
        ...config,
        headers: { 'Content-Type': 'application/json', ...config.headers },
    };

    return fetch(`${API_URL}/${id}`, config).then((resp) => resp.json());
}