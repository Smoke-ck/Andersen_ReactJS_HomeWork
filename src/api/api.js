const API_URL = 'https://fakestoreapi.com/products';

export function getProducts() {
    return request();
}

export function getProduct(id) {
    return request(
        {
            method: 'GET',
        },
        id
    );
}
// export function updateProduct(product) {       
//     request(
//         {
//             method: 'PUT',
//             body: JSON.stringify(product),
//         },
//         product.id,
//     );
// }   --I don't know why it doesn't work--

export const updateProduct = async (product) => {
    const options = {
        method: "PUT",
        body: JSON.stringify(product),
        headers: {
            "Content-Type": "application/json"
        }
    };

    const response = await fetch(`${API_URL}/${product.id}`, options);
    const result = await response.json();

    return result;
}

export const request = async (config = {}, id = '') => {
    config = {
        method: 'GET',
        ...config,
        headers: { 'Content-Type': 'application/json', ...config },
    };

    const response = await fetch(`${API_URL}/${id}`, config);
    const result = await response.json();

    return result;
}


