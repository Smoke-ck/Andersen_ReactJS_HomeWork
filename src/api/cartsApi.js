const API_URL = 'https://fakestoreapi.com/carts';
const API_URL_USER = 'https://fakestoreapi.com/carts/user';

export function getCarts() {
    return request(API_URL_USER, {}, 3);
}

export function addProductInCart(cartData) {
    return request(API_URL, {

        method: "POST",
        body: JSON.stringify({
            userId: 3,
            date: '2020-01-03',
            products: [{
                productId: cartData.id,
            }]
        }),
    })
}

export function updateProductInCart(cartData) {
    return request(API_URL_USER, {

        method: "PUT",
        body: JSON.stringify({
            userId: 3,
            date: '2020-01-03',
            products: [{
                productId: cartData.id,
                quantity: cartData.quantity,
            }]
        }, 3),
    })
}

export function deleteProductInCart(id) {
    return request(API_URL, {
        method: "DELETE",
    }, id)
}

export const request = async (API, config = {}, id = '') => {
    config = {
        method: 'GET',
        ...config,
        headers: { 'Content-Type': 'application/json', ...config.headers },
    };
    const response = await fetch(`${API}/${id}`, config);
    const result = await response.json();

    return result;
}