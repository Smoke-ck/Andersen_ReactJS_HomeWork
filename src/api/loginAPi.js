const API_URL = 'https://fakestoreapi.com/auth/login';
export const loginUser = async (user) => {
  const config = {
    method: "POST",
    body: JSON.stringify({
      username: user.name,
      password: user.password
    }),
    headers: {
      "Content-Type": "application/json"
    }
  };

  const response = await fetch(`${API_URL}`, config);
  const result = await response.json();

  return result;
}