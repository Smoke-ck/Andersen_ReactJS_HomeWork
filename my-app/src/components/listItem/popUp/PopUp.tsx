import React, { FC } from 'react'
import { IToDos } from '../../../api'

type IPopup = {
    handleFavoriteToDo: (e:React.MouseEvent) => void,
    handleDoneToDo: (e:React.MouseEvent) => void,
    handleClick: () => void,
    handleToggleModal: () => void, 
    item: IToDos,
}

const PopUp:FC<IPopup> = ({handleFavoriteToDo,handleDoneToDo,handleClick,handleToggleModal,item}) => {
    return (
        <div>
            <button
                className="listItem__menu--button"
                onClick={(e: React.MouseEvent) => handleFavoriteToDo(e)}>
                {item.favorite ? 'Убрать из избранного ' : 'В избранно'}
            </button>
            <button
                className="listItem__menu--button"
                onClick={(e: React.MouseEvent) => handleDoneToDo(e)}>
                {item.completed ? 'Вернуть в работу' : 'Выполнено'}
            </button>
            <button
                className="listItem__menu--button"
                onClick={handleClick}> Редактировать
            </button>
            <button
                className="listItem__menu--button"
                onClick={handleToggleModal}> Удалить
            </button>
        </div>
    )
}

export default PopUp
