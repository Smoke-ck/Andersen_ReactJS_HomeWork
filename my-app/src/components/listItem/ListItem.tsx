import React, { FC } from "react";
import { IToDos } from "../../api";
import { useState } from "react";
import Modal from "./../modal/Modal";
import { FaStar } from 'react-icons/fa'
import './ListItem.scss'
import PopUp from "./popUp/PopUp";

type IToDoListItem = {
    item: IToDos,
    onItemToggle: (id: number | string) => Promise<IToDos>,
    onItemDelete: (id: number | string) => void,
    onItemFavorite: (id: number | string) => Promise<IToDos>,
    onUpdate: (id: number | string, title: string) => Promise<IToDos>
}

const ListItem: FC<IToDoListItem> = ({ item, onItemToggle, onItemDelete, onItemFavorite, onUpdate }) => {

    const [isEditing, setIsEditing] = useState(false);

    const [title, setTitle] = useState(item.title);

    const [isModalActive, setModalActive] = useState(false);

    const [isDetailsActive, setDetailsActive] = useState(false);
    const [erorr, setNewError] = useState('')

    const handleClick = () => {
        setIsEditing(true);
    }

    const handleKeyPress = (e: React.KeyboardEvent<HTMLDivElement>) => {
        if (e.key === "Enter" && title.length < 160) {
            setIsEditing(false);
            onUpdate(item.id, title.trim())
        }
    }
    const onValueChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTitle(e.target.value)
        if (e.target.value.length > 160) {
            setNewError(`${e.target.value.length - 160}`)
        }
    }
    const handleDetailsActive = () => {
        setDetailsActive(!isDetailsActive);
    }

    const handleFavoriteToDo = (e: React.MouseEvent) => {
        handleDetailsActive();
        e.stopPropagation();
        onItemFavorite(item.id);
    }

    const handleDoneToDo = (e: React.MouseEvent) => {
        handleDetailsActive();
        e.stopPropagation();
        onItemToggle(item.id);
    }

    const handleTogleStar = (e: React.MouseEvent) => {
        e.stopPropagation();
        onItemFavorite(item.id)
    }

    const handleCloseModal = () => {
        setModalActive(false);
    }

    const handleToggleModal = () => {
        setModalActive(prevState => !prevState);
    }

    const handleDeleteToDo = (e: React.MouseEvent) => {
        e.stopPropagation();
        onItemDelete(item.id);
    }

    const handleCloseMenu = (e: React.MouseEvent) => {
        setDetailsActive(false);
        e.stopPropagation();
    }

    return (
        <>
            {isEditing
                ? <div><input
                    className="form__input"
                    value={title}
                    onChange={onValueChange}
                    autoFocus
                    onKeyPress={handleKeyPress}
                    type="text" />
                    {title.length > 160 ? <span className="errorMsg">Превышен лимит текста задачи на {erorr} символов</span> : ''}
                </div>
                :
                <div className="listItem__wrapper">
                    <span onClick={handleDetailsActive} className="listItem__menu--title">...</span>
                    <div
                        onClick={(e: React.MouseEvent) => handleCloseMenu(e)}
                        className={"listItem" + (item.completed ? 'done' : '')}>
                        <div>
                            <div>
                                <h2 className="listItem__title">{item.title} </h2>
                            </div>
                            <div className="listItem__menu">
                                {isDetailsActive
                                    ?
                                    <PopUp
                                        handleFavoriteToDo={handleFavoriteToDo}
                                        handleDoneToDo={handleDoneToDo}
                                        handleClick={handleClick}
                                        handleToggleModal={handleToggleModal}
                                        item={item} />
                                    : ''}
                            </div>
                        </div>
                        <div
                            className={!item.favorite ? "listItem___button--favorite" : 'listItem__button--notFavorite'}
                            onClick={(e: React.MouseEvent) => handleTogleStar(e)}> <FaStar/>
                        </div>
                        {isModalActive && (<Modal
                            active={isModalActive}
                            setActive={setModalActive}
                            handleToggleModal={handleToggleModal}>
                            <div className="deleteMenu">
                                <h3 className="deleteMenu__title">Вы действительно хотите удалить эту задачу?</h3>
                                <div className="deleteMenu__title--todo">{title}</div>
                                <div >
                                    <button className="deleteMenu__button deleteMenu__button--cancel"
                                        onClick={handleCloseModal}>Отмена
                                    </button>
                                    <button className="deleteMenu__button deleteMenu__button--ok"
                                        onClick={(e: React.MouseEvent) => { handleDeleteToDo(e) }}>Да, удалить
                                    </button>
                                </div>
                            </div>
                        </Modal>)}
                    </div>
                </div>
            }
        </>
    );
}

export default ListItem;