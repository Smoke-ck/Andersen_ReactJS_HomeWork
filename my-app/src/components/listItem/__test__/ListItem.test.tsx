import ListItem, { } from "../ListItem"
import { shallow, render, mount } from "enzyme";
import { IToDos } from "../../../api";
import { shallowToJson } from 'enzyme-to-json';
import PopUp from "../popUp/PopUp";

const props = {
    item: {
        title: "Hi from test",
        completed: false,
        favorite: false,
        id: 50,
    },
    onItemToggle: jest.fn(),
    onItemDelete: jest.fn(),
    onItemFavorite: jest.fn(),
    onUpdate: jest.fn(),
}

const event = Object.assign(jest.fn(), {stopPropagation: () => {}})

describe('ListItem', () => {
    let wrapper: any;

    beforeEach(() => {
        wrapper = mount(<ListItem {...props} />)
    })
    
    it("shoud render wrapper", () => {
       const click= wrapper.find('[className="listItem__wrapper"]');
       expect(click.length).toBe(1);
       expect(shallowToJson(wrapper)).toMatchSnapshot()
        
    })
    it("shoud toggle favorite todo", () => { 
        wrapper.find('[className="listItem___button--favorite"]').simulate('click',event);
        expect(props.onItemFavorite).toBeCalled();
    })
    it("shoud toggle favorite todo", () => { 
        wrapper.find('[className="listItem__wrapper"]').simulate('click',event);
        expect(wrapper.find(PopUp)).toBeTruthy()
    })
})
