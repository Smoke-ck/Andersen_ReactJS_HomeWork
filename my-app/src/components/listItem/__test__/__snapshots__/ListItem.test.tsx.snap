// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem shoud render wrapper 1`] = `
<ListItem
  item={
    Object {
      "completed": false,
      "favorite": false,
      "id": 50,
      "title": "Hi from test",
    }
  }
  onItemDelete={[MockFunction]}
  onItemFavorite={[MockFunction]}
  onItemToggle={[MockFunction]}
  onUpdate={[MockFunction]}
>
  <div
    className="listItem__wrapper"
  >
    <span
      className="listItem__menu--title"
      onClick={[Function]}
    >
      ...
    </span>
    <div
      className="listItem"
      onClick={[Function]}
    >
      <div>
        <div>
          <h2
            className="listItem__title"
          >
            Hi from test
             
          </h2>
        </div>
        <div
          className="listItem__menu"
        />
      </div>
      <div
        className="listItem___button--favorite"
        onClick={[Function]}
      >
         
        <FaStar>
          <IconBase
            attr={
              Object {
                "viewBox": "0 0 576 512",
              }
            }
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                key="0"
              />
            </svg>
          </IconBase>
        </FaStar>
      </div>
    </div>
  </div>
</ListItem>
`;
